### PR TITLE
Adding challenge based auth support for keyvault clients

### PR DESCRIPTION
**Pull request by schaabs**
Initial implementation of Challenge based authentication for key vault clients
- Will update to support all clients
- Tests still needed

### CODE REVIEW

File: sdk/keyvault/Azure.Security.KeyVault.Secrets/src/Azure.Security.KeyVault.Secrets.csproj [hunk 1]
```diff
@@ -26,5 +26,6 @@
   <ItemGroup>
     <Compile Include="$(AzureCoreSharedSources)PageResponseEnumerator.cs" />
     <Compile Include="$(AzureCoreSharedSources)ArrayBufferWriter.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)..\..\Shared\ChallengeBasedAuthenticationPolicy.cs" />
```
Comment (pakrym) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets/src/Azure.Security.KeyVault.Secrets.csproj]: I don't think you need `$(MSBuildThisFileDirectory)` Includes are relative to project directory by default

File: sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+using Azure.Core;
+using Azure.Core.Http;
+using Azure.Core.Pipeline;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Azure.Security.KeyVault
+{
+    internal class ChallengeBasedAuthenticationPolicy : HttpPipelinePolicy
```
Comment (pakrym) [inline sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs]: Needs dedicated tests.

File: sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+            // if we get a 401
+            if (message.Response.Status == 401)
+            {
+                // set the content to the original content in case it was cleared
+                message.Request.Content = originalContent;
+
+                // update the cached challenge
+                var challenge = AuthenticationChallenge.GetChallenge(message);
+
+                if (_challenge == null || challenge != _challenge)
```
Comment (pakrym) [inline sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs]: `challenge != _challenge` is a reference comparison, is it intended?

File: sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+            {
+                if (base.Equals(obj))
+                {
+                    return true;
+                }
+
+                AuthenticationChallenge other = obj as AuthenticationChallenge;
+
+                // This assumes that Scopes is always non-null and of length one.  
+                // This is guaranteed by the way the AuthenticationChallenge cache is constructued.
```
Comment (pakrym) [inline sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs]: `Debug.Assert(this.Scopes.Length == 0);`

File: sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+
+                // remove the bearer challenge prefix
+                var trimmedChallenge = challengeValue.Substring(BearerChallengePrefix.Length + 1);
+
+                // Split the trimmed challenge into a set of name=value strings that
+                // are comma separated. The value fields are expected to be within
+                // quotation characters that are stripped here.
+                String[] pairs = trimmedChallenge.Split(new String[] { "," }, StringSplitOptions.RemoveEmptyEntries);
+
+                if (pairs != null && pairs.Length > 0)
```
Comment (pakrym) [inline sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs]: pairs would never be null

Comment (pakrym) [reply sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs]: nit: `string`

File: sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+                // update the cached challenge
+                var challenge = AuthenticationChallenge.GetChallenge(message);
+
+                if (_challenge == null || challenge != _challenge)
+                {
+                    // update the cached challenge
+                    _challenge = challenge;
+
+                    // if no auth challenge was returned raise an exception since we don't know scopes to authenticate
+                    if (_challenge == null)
```
Comment (pakrym) [inline sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs]: Can we bring this condition out and check it right after `var challenge = AuthenticationChallenge.GetChallenge(message);` or does `f (_challenge == null || challenge != _challenge)` matter here too?

File: sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs (hunk 3 above)
Comment (pakrym) [reply sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs]: What is the intended behavior if there was no _challenge and none is received? Flowing the 401 up to the user I suspect?

File: sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+                _refreshOn = token.ExpiresOn - TimeSpan.FromMinutes(2);
+            }
+
+            message.Request.Headers.SetValue(HttpHeader.Names.Authorization, _headerValue);
+        }
+
+        internal class AuthenticationChallenge
+        {
+            private static readonly Dictionary<string, AuthenticationChallenge> _cache = new Dictionary<string, AuthenticationChallenge>();
+            private static readonly object _cacheLock = new object();
```
Comment (pakrym) [inline sdk/keyvault/Shared/ChallengeBasedAuthenticationPolicy.cs]: You can lock on the `_cache` itself.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/tests/KeyClientLiveTests.cs [hunk 8]
```diff
@@ -19,6 +19,19 @@ public KeyClientLiveTests(bool isAsync) : base(isAsync)
...
+        [SetUp]
+        public void ClearChallengeCacheforRecord()
+        {
+            // in record mode we reset the challenge cache before each test so that the challenge call
+            // is always made.  This allows tests to be replayed independently and in any order
+            if (this.Mode == RecordedTestMode.Record || this.Mode == RecordedTestMode.Playback)
+            {
+                this.Client = this.GetClient();
+
+                ChallengeBasedAuthenticationPolicy.AuthenticationChallenge.ClearCache();
```
Comment (pakrym) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/tests/KeyClientLiveTests.cs]: I wonder if it's worth to have this as an internal client option (something like IsGlobalChallengeCacheEnabled) and set it in `GetClient`, not sure if it's cleaner though.

**Comment by pakrym** [comment]
File an issue for adding tests.

**Review by pakrym — APPROVED** [review]

**Comment by schaabs** [comment]
Added issue https://github.com/Azure/azure-sdk-for-net/issues/7219 to track adding unit tests for ChallengeBasedAuthenticationPolicy

**Pull request merged**